fortuna/hallucination/base.py

```diff
@@ -0,0 +1,335 @@
+from typing import (
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+from sklearn.manifold import locally_linear_embedding
+from sklearn.mixture import GaussianMixture
+import torch
+from torch import nn
+from tqdm import tqdm
+from transformers import PreTrainedTokenizer
+
+from fortuna.conformal import BinaryClassificationMulticalibrator
+from fortuna.hallucination.grouping.clustering.base import GroupingModel
+from fortuna.hallucination.scoring.inv_perplexity import inv_perplexity
+
+
+class HallucinationMulticalibrator:
+    def __init__(
+        self,
+        generative_model: nn.Module,
+        tokenizer: PreTrainedTokenizer,
+        embedding_reduction_fn: Callable[[np.ndarray], np.ndarray] = None,
+        clustering_models: Optional[List] = None,
+        scoring_fn: Callable[
+            [torch.Tensor, torch.Tensor, int], torch.Tensor
+        ] = inv_perplexity,
+    ):
+        """
+        A hallucination multicalibrator class.
+        Given some context (e.g. a question), some text (e.g. one or multiple answers) and some scoring function,
+        it multicalibrates the probability that the text is a hallucination or not based on a supervised calibration
+        dataset.
+        Under the hood, the method constructs model-based low-dimensional embeddings of the initial contexts plus texts,
+        it clusters them to provide meaningful (possibly overlapping) groups,
+        it computes the scores for each text,
+        and it finally calibrates the scores on each subset formed by the intersection of each group and each level set
+        of the score function.
+
+        Parameters
+        ----------
+        generative_model: nn.Module
+            A generative model.
+        tokenizer: PreTrainedTokenizer
+            A tokenizer.
+        embedding_reduction_fn: Callable[[np.ndarray], np.ndarray]
+            A function aimed at reducing the embedding dimensionality.
+        clustering_models: Optional[List]
+            A list of clustering models.
+        scoring_fn: Callable[[torch.Tensor, torch.Tensor, int], torch.Tensor]
+            A scoring function.
+        """
+        self.generative_model = generative_model
+        self.tokenizer = tokenizer
+        if embedding_reduction_fn is not None:
+            self.embedding_reduction_fn = embedding_reduction_fn
+        else:
+            self.embedding_reduction_fn = locally_linear_embedding_fn
```

Comment:
`self._embedding_reduction_fn = embedding_reduction_fn or locally_linear_embedding_fn` if you like it